swiftsimio/initial_conditions/generate_particles.py

```diff
@@ -1,5 +1,1304 @@
+    (https://arxiv.org/abs/1907.11250).
+
+    Attributes
+    -------------
+
+    self.rhofunc: callable
+        The density function that is to be reproduced in the initial conditions.
+        It must take two positional arguments:
+        
+        - ``x``: np.ndarray of 3D particle coordinates (even if your initial 
+          conditions have lower dimensionality)
+        - ``ndim``: integer, number of dimensions that your simulations is to have
+
+    self.boxsize: unyt.unyt_array
+        The box size of the simulation.
+
+    self.nx: int
+        how many particles along every dimension you want your simulation
+        to contain
+
+    self.ndim: int
+        how many dimensions you want your simulation to have
+
+    self.unitsys: unyt.unit_systems.UnitSystem
+        a unit system that contains the units you want your ICs to have
+
+    self.periodic: bool
+        whether the simulation box is periodic or not
+
+    self.kernel: str {'cubic spline',}
+        which kernel to use
+
+    self.eta: float, optional
+        resolution eta, which defines the number of neighbours used
+        independently of dimensionality
+
+    self.rho_max: float or None
+        The maximal density within the simulation box. If ``None``, an 
+        approximate value will be determined if the rejection sampling to obtain
+        an initial particle configuration guess is used.
+
+    self.runparams: RunParams
+        A ``RunParams`` instance that contains runtime parameters for the generator.
+        The user is encouraged to tinker with these parameters before the iteration
+        is commenced.
+
+    self.iterparams: IterData
+        An ``IterData`` instance that holds data relevant for iteration steps. The
+        user is *strongly discouraged* of fiddling with this.
+
+    self.stats: IterStats
+        An ``IterStats`` instance that holds iteration statistics that should be
+        accessible after the iteration finished.
+
+    self.npart: int
+        total number of particles in simulation
+
+    self.coordinates: unyt.unyt_array
+        unyt_array containing the final particle coordinates with shape (self.npart, 3).
+
+    self.masses: unyt.unyt_array
+        unyt_array containing the final particle masses with shape (self.npart).
+
+    self.smoothing_lengths: unyt.unyt_array
+        unyt_array containing the final particle smoothing lengths with shape (self.npart, 3).
+
+    self.densities: unyt.unyt_array
+        unyt_array containing the final particle densities with shape (self.npart).
+
+    self.x: np.ndarray
+        unitless particle coordinates that will be worked with/on with shape (self.npart, 3).
+
+    self.m: np.ndarray
+        unitless particle masses that will be worked with/on with shape (self.npart).
+
+
+
+
+    Notes
+    -----------
+    
+    + The default workflow should be something like
+        ``pg = ParticleGenerator(...)``
```

Comment:
You should use a `code-block` here like:

```
Notes
-----

The default workflow looks something like:

.. code-block:: python

    pg = ParticleGenerator()

    pg.runparams.iter_max = 10000
    pg.runparams.whatever_parameter = ...

    pg.initial_setup()

    pg.run_iteration()

Reply:
Done.